### PR TITLE
Memprof: provide the guarantee that an allocation callback is always run in the same thread the allocation takes place.

### PR DESCRIPTION
**Pull request by jhjourdan**
This is done by maintaining a bag of pending allocation callbacks for each thread, in the form of an linked list.

We also change the behavior of `Memprof.stop`, so that it does no longer try to run pending callbacks. This is only effective in native mode, since function calls in bytecode mode will trigger polling and hence run pending callbacks.

This MR is rather large, but only one commit actually has non-trivial content. Reviewing should be easier by reading one commit after the other.

### CODE REVIEW

**Comment by jhjourdan** [comment]
> We also change the behavior of Memprof.stop, so that it does no longer try to run pending callbacks. This is only effective in native mode, since function calls in bytecode mode will trigger polling and hence run pending callbacks.

Perhaps I could say a bit more on that: first, this is something which have been discussed several times (including in #8920), and second, in a multithreaded setting, it makes little sense to try to handle the callbacks of the current thread while discarding the others.

File: runtime/memprof.c [hunk 1]
```diff
@@ -939,3 +939,26 @@ CAMLprim value caml_memprof_stop(value unit)
...
+void caml_memprof_init_th_ctx(struct caml_memprof_th_ctx* ctx) {
+  ctx->suspended = 0;
+}
+
+void caml_memprof_stop_th_ctx(struct caml_memprof_th_ctx* ctx) {
+  /* Make sure that no memprof callback is being executed in this
+     thread. If so, memprof data structures may have pointers to the
+     thread's stack. */
+  if(ctx->suspended)
+    caml_fatal_error("Thread.exit called from a memprof callback.");
```
Comment (gadmm) [inline runtime/memprof.c:1096]: Is the call to `caml_fatal_error` necessary here, or would it be possible to raise an exception instead? For instance `caml_thread_exit` already uses `caml_invalid_argument`.

File: runtime/memprof.c [hunk 2]
```diff
@@ -939,3 +939,26 @@ CAMLprim value caml_memprof_stop(value unit)
...
+
+void caml_memprof_init_th_ctx(struct caml_memprof_th_ctx* ctx) {
+  ctx->suspended = 0;
+}
+
+void caml_memprof_stop_th_ctx(struct caml_memprof_th_ctx* ctx) {
+  /* Make sure that no memprof callback is being executed in this
+     thread. If so, memprof data structures may have pointers to the
+     thread's stack. */
+  if(ctx->suspended)
```
Comment (gadmm) [inline runtime/memprof.c]: This can also be set in `caml_fatal_uncaught_exception`, in that case the error message would be misleading.

The use of suspended in `caml_fatal_uncaught_exception` is replaced by a proper mask at https://github.com/ocaml/ocaml/pull/8961/files#diff-7daa060567fae8578a3e27866a93d610R143 so the problem fixes itself eventually, but not immediately. Do you have an idea of something that could be done in the meantime?

**Comment by stedolan** [comment]
Over in #8920, when responding to @gadmm, I said:

> Making this guarantee can cause unbounded delays. Suppose an allocation is sampled on a thread in a C call (where the callback cannot run promptly), and the C call then enters a blocking section. It can be a very long time before the blocking section returns.

@jhjourdan has since pointed out that this already happens for single-threaded programs, for the same reason. So, making this guarantee will not in fact make anything worse, and it would be nice to be able to rely on it.

**Review by gadmm — COMMENTED** [review]
Here is for the first three commits. Overall they look good to me but I have found two issues in the form of questions. I do the other three separately since you have updated them since I started the review.

**Comment by stedolan** [comment]
@jhjourdan How important is it to get this into 4.11? I won't be able to review it tomorrow, but I could have a look on Monday if it's urgent.

(On the other hand, if you want this in 4.11 because you think the existing behaviour is broken, perhaps we could argue that this is a bugfix and spend a little longer reviewing it)

**Comment by gasche** [comment]
(Personally I think that rushing to review a complex PR a couple days before a deadline is an anti-pattern of collaborative software development.)

**Comment by gadmm** [comment]
Would you like to submit the first patch separately to have it in 4.11? And some of the simple refactorings if you want, they do not look risky.

File: runtime/memprof.c (hunk 1 above)
Comment (jhjourdan) [reply runtime/memprof.c:1096]: Technically, I could make it an exception, but the implementation would be more complex : indeed, at this point, we have already called `caml_thread_save_runtime_state` in `st_stub.c`, so we would either have to do this check earlier or do some kind of rollback. Given that this exception is anyway unlikely to be caught, I think it is fine as is.

File: runtime/memprof.c (hunk 2 above)
Comment (jhjourdan) [reply runtime/memprof.c]: Good catch!

I don't think there is a simple solution (except using two flags, which is ugly). Given that 1- the problem will be fixed eventually and that 2- I don't think anyone will actually call `Thread.exit` from an uncaught exception handler, I think we could let this as is.

Comment (jhjourdan) [reply runtime/memprof.c]: Perhaps we could rename `suspended` into e.g., `callback_running`, to make the purpose of this variable clearer, and add a commentary in `caml_fatal_uncaught_exception` to make it clear that this should go away with masking. What do you think ?

**Comment by jhjourdan** [comment]
@stedolan : don't rush reviewing this for Monday. This is a rather complicated patch, for a feature which I think is relevant but not essential.

That said, I agree with @gadmm that we could consider merging the first commit before 4.11. But it is trivial and could be considered as a bugfix.

**Comment by jhjourdan** [comment]
I moved the commit about the fatal error in `Thread.exit` in #9458 
This PR now depends on #9458.

File: runtime/memprof.c (hunk 2 above)
Comment (jhjourdan) [reply runtime/memprof.c]: Fixed in #9458.

File: runtime/memprof.c (hunk 1 above)
Comment (jhjourdan) [reply runtime/memprof.c:1096]: If you are still concerned about that, please complain in #9458.

Comment (gadmm) [reply runtime/memprof.c:1096]: ok thanks

**Comment by jhjourdan** [comment]
Now that #9458 is merged, I rebased on top of trunk.

File: runtime/memprof.c [hunk 3]
```diff
@@ -464,17 +520,20 @@ static void flush_deleted(void)
       j++;
     }
     i++;
-    if (trackst.young == i) trackst.young = j;
-    if (trackst.callback == i) trackst.callback = j;
+    if (trackst.young_idx == i) trackst.young_idx = j;
+    if (trackst.promote_dealloc_callback_idx == i)
+      trackst.promote_dealloc_callback_idx = j;
   }
```
Comment (stedolan) [inline runtime/memprof.c:594]: I don't see how the next_alloc_callback_idx fields are maintained when everything shuffles around here

**Comment by stedolan** [comment]
I've started reading this, but honestly I find all of the `idx_ptr` manipulation and the mixture of linked lists and array structures very hard to follow. I think we might have the wrong data structure for this problem.

File: runtime/memprof.c (hunk 3 above)
Comment (jhjourdan) [reply runtime/memprof.c:594]: Indeed, this is completely bogus. This is not because of `next_alloc_callback_idx`, but rather because of the idx_ptr pointers which are not updated accordingly.

**Comment by jhjourdan** [comment]
Now that I look at what I have written again, I tend to agree with you :(.

We should probably redesign the data structures. Fortunately, if we get something where each thread allocates in its own data structure, then this will be easier to port to Multicore with good efficiency.

However, I know how much time we needed to spend on the current data structure, and I am afraid that this will take long too... Let me think more about this.

**Comment by jhjourdan** [comment]
Closing this PR in favor of #9674.
